### ISSUE TITLE
bazel version Conflict when build tensorflow/servering with recommender-addons

### ISSUE THREAD

**Issue by uzhy1987** [tensorflow/recommenders-addons#452]
the latest version of recommenders-addons (v0.7.2) use Bazel 5.1.1,  while the tensorflow/serving(v2.15.1) use Bazel 6.4.0,  how can I build a tensorflow/serving with recommenders-addons ops

**Comment by uzhy1987** [comment]
I try the method [https://github.com/tensorflow/tensorflow/issues/58980#issuecomment-1549617092](url) , but it not work. 

I build recommenders-addons (v0.7.2)  and copy _cuckoo_hashtable_ops.so to tensorflow/serving:2.15.1 docker image, and preload the so libs, as shown bellow, but it fail to exec

LD_PRELOAD="/usr/local/lib/libtensorflow_framework.so.2  /usr/local/lib/_cuckoo_hashtable_ops.so /usr/local/lib/_math_ops.so" tensorflow_model_server --port=8500 --rest_api_port=8501 --model_name=${MODEL_NAME} --model_base_path=${MODEL_BASE_PATH}/${MODEL_NAME} "$@"

**Comment by alykhantejani** [comment]
@uzhy1987 please update if you find anything as I am also stuck here

**Comment by uzhy1987** [comment]
I try to build tensorflow/serving 2.11.1 with tfra 0.7.2, they both use Bazel 5.1.1.  But I got failure when linking.  

`ERROR: /data/tfra/serving/tensorflow_serving/model_servers/BUILD:451:10: Linking tensorflow_serving/model_servers/tensorflow_model_server failed: (Exit 1): gcc failed: error executing command
  (cd /data/tfra/serving/.cache/_bazel_root/712f919a74d47ea3310a3e9710065a13/execroot/tf_serving && \
  exec env - \
    FOR_TF_SERVING=1 \
    PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin \
    PWD=/proc/self/cwd \
    TF_CXX11_ABI_FLAG=1 \
    TF_CXX_STANDARD=c++17 \
    TF_HEADER_DIR=/usr/local/lib/python3.9/dist-packages/tensorflow/include \
    TF_SHARED_LIBRARY_DIR=/usr/local/lib/python3.9/dist-packages/tensorflow \
    TF_SHARED_LIBRARY_NAME=libtensorflow_framework.so.2 \
    TF_VERSION_INTEGER=2111 \
  /usr/bin/gcc @bazel-out/k8-opt/bin/tensorflow_serving/model_servers/tensorflow_model_server-2.params)
# Configuration: ff3bf0981d860dc7c7e582b1e32d6098497b1cf4aa66e54f51c115debaf89daa
# Execution platform: @local_execution_config_platform//:platform
bazel-out/k8-opt/bin/tensorflow_recommenders_addons/dynamic_embedding/core/_objs/_cuckoo_hashtable_ops.so/cuckoo_hashtable_op.o:cuckoo_hashtable_op.cc:function tensorflow::lookup::LookupInterface::DebugSt
ring[abi:cxx11]() const: error: undefined reference to 'tsl::strings::StrCat[abi:cxx11](tsl::strings::AlphaNum const&, tsl::strings::AlphaNum const&)'
bazel-out/k8-opt/bin/tensorflow_recommenders_addons/dynamic_embedding/core/_objs/_cuckoo_hashtable_ops.so/cuckoo_hashtable_op.o:cuckoo_hashtable_op.cc:function tsl::Status tsl::errors::InvalidArgument<cha
r const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >): error: undefined
reference to 'tsl::strings::StrCat[abi:cxx11](tsl::strings::AlphaNum const&, tsl::strings::AlphaNum const&)'
bazel-out/k8-opt/bin/tensorflow_recommenders_addons/dynamic_embedding/core/_objs/_cuckoo_hashtable_ops.so/cuckoo_hashtable_op.o:cuckoo_hashtable_op.cc:function std::__cxx11::basic_string<char, std::char_t
raits<char>, std::allocator<char> >* tsl::internal::MakeCheckOpString<int, int>(int const&, int const&, char const*): error: undefined reference to 'tsl::internal::CheckOpMessageBuilder::NewString[abi:cxx
11]()'
bazel-out/k8-opt/bin/tensorflow_recommenders_addons/dynamic_embedding/core/_objs/_cuckoo_hashtable_ops.so/cuckoo_hashtable_op.o:cuckoo_hashtable_op.cc:function std::__cxx11::basic_string<char, std::char_t
raits<char>, std::allocator<char> >* tsl::internal::MakeCheckOpString<unsigned long, unsigned long>(unsigned long const&, unsigned long const&, char const*): error: undefined reference to 'tsl::internal::
CheckOpMessageBuilder::NewString[abi:cxx11]()'
bazel-out/k8-opt/bin/tensorflow_recommenders_addons/dynamic_embedding/core/_objs/_cuckoo_hashtable_ops.so/cuckoo_hashtable_op.o:cuckoo_hashtable_op.cc:function tsl::Status tensorflow::ResourceMgr::LookupI
nternal<tensorflow::lookup::LookupInterface, false>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::al
locator<char> > const&, tensorflow::lookup::LookupInterface**) const: error: undefined reference to 'tensorflow::ResourceMgr::DoLookup(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocat
or<char> > const&, tensorflow::TypeIndex, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, tensorflow::ResourceBase**) const'
bazel-out/k8-opt/bin/tensorflow_recommenders_addons/dynamic_embedding/core/_objs/_cuckoo_hashtable_ops.so/cuckoo_hashtable_op.o:cuckoo_hashtable_op.cc:function tensorflow::recommenders_addons::HashTableOp
Kernel::GetTable(tensorflow::OpKernelContext*, tensorflow::lookup::LookupInterface**): error: undefined reference to 'tensorflow::TensorShapeRep::DebugString[abi:cxx11]() const'
bazel-out/k8-opt/bin/tensorflow_recommenders_addons/dynamic_embedding/core/_objs/_cuckoo_hashtable_ops.so/cuckoo_hashtable_op.o:cuckoo_hashtable_op.cc:function tsl::Status tensorflow::OpKernelConstruction
::GetAttr<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::basic_string_view<char, std::char_traits<char> >, std::__cxx11::basic_string<char, std::char_traits<char>,
std::allocator<char> >*) const: error: undefined reference to 'tensorflow::GetNodeAttr(tensorflow::AttrSlice const&, std::basic_string_view<char, std::char_traits<char> >, std::__cxx11::basic_string<char,
 std::char_traits<char>, std::allocator<char> >*)'`

**Comment by alykhantejani** [comment]
Ok so this is what im doing:

```
docker run -it --network=host --entrypoint /bin/bash tensorflow/serving:2.11.1-devel

cd /home
git clone https://github.com/tensorflow/recommenders-addons.git
cd recommenders-addons
python configure.py
cp .bazelrc tools/serving_padding/.bazelrc_padding

export SERVING_WITH_GPU=0 
export TFRA_BRANCH="master"
export TF_SERVING_BRANCH="r2.11.1"

export TFRA_SERVING_WORKSPACE=/home/
mkdir -p $TFRA_SERVING_WORKSPACE && cd $TFRA_SERVING_WORKSPACE
git clone -b $TF_SERVING_BRANCH https://github.com/tensorflow/serving.git

cd $TFRA_SERVING_WORKSPACE/recommenders-addons/tools
bash config_tfserving.sh $TFRA_BRANCH $TFRA_SERVING_WORKSPACE/serving $SERVING_WITH_GPU
cd $TFRA_SERVING_WORKSPACE/serving

 bazel build tensorflow_recommenders_addons/dynamic_embedding/core:_cuckoo_hashtable_ops.so
```

Because this is giving me an error:

```
ERROR: Traceback (most recent call last):
	File "/home/serving/tensorflow_recommenders_addons/tensorflow_recommenders_addons.bzl", line 10, column 6, in <toplevel>
		"cuda_is_configured",
```

however if I run this build command from `../recommender-addons` it builds. Which leads me to think its a workspace issue with tf serving. 

can you share your WORKSPACE file?

Also maybe @rhdong might be able to assist?

**Comment by alykhantejani** [comment]
ok well if I remove `cuda_is_configured` it seems to work. But this feels hacky (but I am building without GPU)

**Comment by alykhantejani** [comment]
@MoFHeka perhaps you could help?

**Comment by MoFHeka** [comment]
Try to delete all Windows-related code in the bazel file, such as tensorflow_recommenders_addons/tensorflow_recommenders_addons.bzl:44

**Comment by uzhy1987** [comment]
I use the docker image tensorflow/serving:2.11.1-devel, tensorflow/serving 2.11.1 with tfra 0.7.2. 
I change .bazelrc_padding and replace cuda_is_configured() with if_cuda_is_configured(cuda_deps).  

docker run -it --network=host -v /data/tfra:/data/tfra -v /tmp:/tmp tensorflow/serving:2.11.1-devel /bin/bash
cd /data/tfra/serving; TEST_TMPDIR=.cache bazel build tensorflow_serving/model_servers:tensorflow_model_server

the .bazelrc_padding change as follow:
build --action_env TF_HEADER_DIR="/usr/local/lib/python3.9/dist-packages/tensorflow/include"
build --action_env TF_SHARED_LIBRARY_DIR="/usr/local/lib/python3.9/dist-packages/tensorflow"
build --action_env TF_SHARED_LIBRARY_NAME="libtensorflow_framework.so.2"
build --action_env TF_CXX11_ABI_FLAG="0"
build --action_env TF_CXX_STANDARD="c++17"
build --action_env TF_VERSION_INTEGER="2111"
build --action_env FOR_TF_SERVING="1"
build --spawn_strategy=standalone
build --strategy=Genrule=standalone
build -c opt
build --copt=-mavx

**Comment by MoFHeka** [comment]
TFRA also support compiling with Bazel 6.4.0 if remove those windows things. So there is no need to downgrade the TF version.

**Comment by alykhantejani** [comment]
Ok So Im doing it with 2.15.1 and bazel 6.4.0. Got it almost working but linking is failing with linking TSL:
```
error: undefined reference to 'tsl::TfCheckOpHelperOutOfLine
``` etc. 

Any ideas? Did you also face this? I have set `TF_CXX11_ABI_FLAG=0`

**Comment by alykhantejani** [comment]
ok turns out my installed tf version wasn't correct. I can send a PR with the bazel fixes to upgrade to 6.4.0
